algorithms/src/std_algorithms/impl/Kokkos_AdjacentDifference.hpp

```diff
@@ -82,9 +82,17 @@ OutputIteratorType adjacent_difference_exespace_impl(
     return first_dest;
   }
 
-  // run
+  // ranges shall not overlap
   const auto num_elements =
       Kokkos::Experimental::distance(first_from, last_from);
+  auto last_dest = first_dest + num_elements;
+  // Iterators must be accessed from Execution space
+  ::Kokkos::parallel_for(
+      "Check", RangePolicy<ExecutionSpace>(ex, 0, 1), KOKKOS_LAMBDA(int) {
+        Impl::expect_no_overlap(first_from, last_from, first_dest, last_dest);
+      });
```

Comment:
```suggestion
#ifdef KOKKOS_ENABLE_DEBUG
  ::Kokkos::parallel_for(
      "Check", RangePolicy<ExecutionSpace>(ex, 0, 1), KOKKOS_LAMBDA(int) {
        Impl::expect_no_overlap(first_from, last_from, first_dest, last_dest);
      });
#endif
```
so we don't have to always pay for the launch overhead for another kernel.

Reply:
That is true. I also add this guard to the team level, since this check performs some computations